Make _WKApplicationManifest initializer return nil when jsonData is not valid JSON

szewai opened the pull request:
#### fa54f7ce3cd9b71732786437b9d5dcde72a4c07f
<pre>
Make _WKApplicationManifest initializer return nil when jsonData is not valid JSON
<a href="https://bugs.webkit.org/show_bug.cgi?id=272161">https://bugs.webkit.org/show_bug.cgi?id=272161</a>
<a href="https://rdar.apple.com/125913241">rdar://125913241</a>

Reviewed by Chris Dumez.

API Test: WKApplicationManifest.InvalidJSONData

* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
(WebCore::ApplicationManifestParser::parse):
(WebCore::ApplicationManifestParser::parseWithValidation):
(WebCore::ApplicationManifestParser::createJSONObject):
(WebCore::ApplicationManifestParser::parseManifest):
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.mm:
(-[_WKApplicationManifest initWithJSONData:manifestURL:documentURL:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ApplicationManifest.mm:
(TestWebKitAPI::TEST(WKApplicationManifest, InvalidJSONData)):

Canonical link: <a href="https://commits.webkit.org/277098@main">https://commits.webkit.org/277098@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/92ce77d80c9c950b6a1a2db4e243e188ed2bcc64

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46673 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25833 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49292 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49350 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42720 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48980 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30193 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23295 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/38042 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47253 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22814 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40203 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19313 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/20188 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41333 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4718 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42952 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41703 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51206 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21682 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/18063 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/45308 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/22970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44262 "Found 1 new API test failure: /WebKitGTK/TestWebViewEditor:/webkit/WebKitWebView/cut-copy-paste/non-editable (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10318 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/23429 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22674 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
```diff
@@ -50,29 +50,47 @@ ApplicationManifest ApplicationManifestParser::parse(const String& source, const
 }
 
-ApplicationManifest ApplicationManifestParser::parseManifest(const String& text, const URL& manifestURL, const URL& documentURL)
+ApplicationManifest ApplicationManifestParser::parseManifest(const String& text, const URL& manifestURL, const URL& documentURL, bool shouldFinishOnError)
```
This `shouldFinishOnError` boolean parameter isn't nice. Also, ideally, we wouldn't construct an `ApplicationManifest` at all when JSON parsing fails. My suggestion would be:
1. Move the logic to convert the String to a JSObject (e.g. do the JSON parsing) to a new utility function and call it at the call sites of parseManifest()
2. Rename parseManifest() to constructManifest() and have it take the parsed JSON result as input
3. Your new parseWithValidation() function only needs to call the parsing function but never attempt to construct an ApplicationManifest.

szewai replied:
Implemented the suggestion in the new patch, except renaming `parseManifest` since all functions in the class that takes `JSON::Object& manifest` are using `parse` as prefix.
But let me know if you still prefer `constructManifest ` and I will change it.

cdumez reviewed (approved):
Yes, I think this looks better. Thanks for making the changes.

szewai committed:
Make _WKApplicationManifest initializer return nil when jsonData is not valid JSON
https://bugs.webkit.org/show_bug.cgi?id=272161
rdar://125913241

Reviewed by Chris Dumez.

API Test: WKApplicationManifest.InvalidJSONData

* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
(WebCore::ApplicationManifestParser::parse):
(WebCore::ApplicationManifestParser::parseWithValidation):
(WebCore::ApplicationManifestParser::createJSONObject):
(WebCore::ApplicationManifestParser::parseManifest):
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.mm:
(-[_WKApplicationManifest initWithJSONData:manifestURL:documentURL:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ApplicationManifest.mm:
(TestWebKitAPI::TEST(WKApplicationManifest, InvalidJSONData)):

Canonical link: https://commits.webkit.org/277098@main
